app-2/src/main/java/tech/relaycorp/letro/awala/AwalaManager.kt

```diff
@@ -0,0 +1,235 @@
+                        type = MessageType.from(message.type),
+                        content = message.content,
+                    ).also { Log.d(TAG, "Receive message: ($it)") }
+                )
+                message.ack()
+            }
+        }
+
+        isReceivingMessages = true
+    }
+
+    private suspend fun configureAwala() {
+        registerFirstPartyEndpointIfNeeded()
+        importServerThirdPartyEndpointIfNeeded()
+        authoriseReceivingMessagesFromThirdPartyEndpointIfNeeded()
+        startReceivingMessages()
+    }
+
+    private suspend fun checkIfAwalaAppInstalled(): Boolean {
+        try {
+            GatewayClient.bind()
+            configureAwala()
+        } catch (exp: GatewayBindingException) {
+            this.isAwalaInstalledOnDevice = false
+            return false
+        }
+        this.isAwalaInstalledOnDevice = true
+        return true
+    }
+
+    private suspend fun registerFirstPartyEndpointIfNeeded(): FirstPartyEndpoint? {
+        if (awalaRepository.getServerFirstPartyEndpointNodeId() != null) {
+            return null
+        }
+        val firstPartyEndpoint = FirstPartyEndpoint.register()
+        awalaRepository.saveServerFirstPartyEndpointNodeId(firstPartyEndpoint.nodeId)
+        return firstPartyEndpoint
+    }
+
+    private suspend fun authoriseReceivingMessagesFromThirdPartyEndpointIfNeeded() {
+        if (awalaRepository.isAuthorizedReceiveMessagesFromServer()) {
+            return
+        }
+        val firstPartyEndpointNodeId = awalaRepository.getServerFirstPartyEndpointNodeId()
+            ?: registerFirstPartyEndpointIfNeeded()?.nodeId
+            ?: ""
+
+        val thirdPartyEndpointNodeId = awalaRepository.getServerThirdPartyEndpointNodeId()
+            ?: importServerThirdPartyEndpointIfNeeded()?.nodeId
+            ?: ""
+
+        if (firstPartyEndpointNodeId.isEmpty() || thirdPartyEndpointNodeId.isEmpty()) {
+            // TODO: log error
+            Log.i(TAG, "authoriseReceivingMessagesFromThirdPartyEndpoint(). Some of ids is empty: $firstPartyEndpointNodeId ; $thirdPartyEndpointNodeId")
+            return
+        }
+
+        val firstPartyEndpoint = loadNonNullFirstPartyEndpoint(firstPartyEndpointNodeId)
+        val thirdPartyEndpoint = loadNonNullThirdPartyEndpoint(thirdPartyEndpointNodeId)
+
+        // Create the Parcel Delivery Authorisation (PDA)
+        val auth = firstPartyEndpoint.authorizeIndefinitely(thirdPartyEndpoint)
+
+        Log.d(TAG, "authoriseReceivingMessagesFromThirdPartyEndpointIfNeeded() ${firstPartyEndpoint.nodeId}:${thirdPartyEndpoint.nodeId}")
+
+        // Send it to the server
+        val authMessage = OutgoingMessage.build(
+            MessageType.AuthorizeReceivingFromServer.value,
+            auth,
+            firstPartyEndpoint,
+            thirdPartyEndpoint,
+        )
+
+        GatewayClient.sendMessage(authMessage)
+
+        awalaRepository.setAuthorizedReceiveMessagesFromServer(true)
```

Comment:
To simplify things, I think we should get rid of `awalaRepository.setAuthorizedReceiveMessagesFromServer` and instead unconditionally send the authorisation in `importServerThirdPartyEndpointIfNeeded()` (when the 3rd party endpoint is imported).

Unless I'm missing something, the only advantage of storing this in the preferences is in the very unlikely event that the app crashes between importing the 3rd party endpoint and issuing the authorisation. However, the approach I'm proposing above would still result in the auth being sent, and it doesn't matter if it's sent multiple times because it's idempotent.

Reply:
Maybe it's excessive, I second that. Thanks for the comment